File: trunk/java/picocontainer/src/java/org/picocontainer/extras/InvokingComponentAdapterFactory.java
package org.picocontainer.extras;

import org.picocontainer.*;
import org.picocontainer.defaults.AssignabilityRegistrationException;
import org.picocontainer.defaults.ComponentAdapterFactory;
import org.picocontainer.defaults.NotConcreteRegistrationException;
import org.picocontainer.defaults.PicoInvocationTargetInitializationException;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * A generic ComponentAdapter that will invoke a particular method
 * on the component after it has been instantiated. This can be useful
 * for implementing simple lifecycle support without the components
 * having to implement any interfaces.
 *
 * @author Aslak Helles&oslash;y
 * @deprecated Use the lifcycle support in PicoContainer instead.
 * @version $Revision$
 */
public class InvokingComponentAdapterFactory extends DecoratingComponentAdapterFactory {
    private String methodName;
    private Class[] parameterTypes;
    private Object[] arguments;

    public InvokingComponentAdapterFactory(ComponentAdapterFactory delegate,
                                           String methodName, Class[] parameterTypes, Object[] arguments) {
        super(delegate);
        this.methodName = methodName;
        this.parameterTypes = parameterTypes;
        this.arguments = arguments;
    }

    public ComponentAdapter createComponentAdapter(Object componentKey, Class componentImplementation, Parameter[] parameters) throws PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
        return new Adapter(super.createComponentAdapter(componentKey, componentImplementation, parameters));
    }

    public class Adapter extends DecoratingComponentAdapter {
        private Method method;
        private Object componentInstance = null;

        public Object getInvocationResult() {
            return invocationResult;
        }

        private Object invocationResult;

        public Adapter(ComponentAdapter delegate) {
            super(delegate);
            try {
                method = delegate.getComponentImplementation().getMethod(methodName, parameterTypes);
            } catch (NoSuchMethodException e) {
                method = null;
            } catch (final SecurityException e) {
                ///CLOVER:OFF
                method = null;
                ///CLOVER:ON                
            }
        }

        public Object getComponentInstance() throws PicoInitializationException, PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
            if (componentInstance == null) {
                componentInstance = super.getComponentInstance();

                if (method != null) {
                    try {
                        invocationResult = method.invoke(componentInstance, arguments);
                    } catch (IllegalAccessException e) {
                        throw new PicoInvocationTargetInitializationException(e);
                    } catch (final IllegalArgumentException e) {
                        throw new PicoInitializationException() {
                            public String getMessage() {
                                return "Illegal argument:" + e.getMessage();
                            }
                        };
                    } catch (InvocationTargetException e) {
                        throw new PicoInvocationTargetInitializationException(e.getTargetException());
                    }
                }
            }
            return componentInstance;
        }
    }
}

File: dotnet/picocontainer/src/Extras/InvokingComponentAdapterFactory.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.html file.                                                    *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer;
using PicoContainer.Defaults;

namespace PicoContainer.Extras
{
	
	public class InvokingComponentAdapterFactory:DecoratingComponentAdapterFactory
	{
		private string methodName;
		private System.Type[] parameterTypes;
		private object[] arguments;
		
		public InvokingComponentAdapterFactory(ComponentAdapterFactory theDelegate, string methodName, System.Type[] parameterTypes, object[] arguments):base(theDelegate)
		{
			this.methodName = methodName;
			this.parameterTypes = parameterTypes;
			this.arguments = arguments;
		}
		
		public override ComponentAdapter CreateComponentAdapter(object componentKey, System.Type componentImplementation, Parameter[] parameters)
		{
			return new Adapter(this,base.CreateComponentAdapter(componentKey, componentImplementation, parameters));
		}
		

		public class Adapter:DecoratingComponentAdapter
		{
			private class AnonymousClassPicoInitializationException : Exception
			{
				public AnonymousClassPicoInitializationException(System.ArgumentException e, Adapter enclosingInstance)
				{
					InitBlock(e, enclosingInstance);
				}
				private void  InitBlock(System.ArgumentException e, Adapter enclosingInstance)
				{
					this.e = e;
					this.enclosingInstance = enclosingInstance;
				}

				private System.ArgumentException e;
				private Adapter enclosingInstance;
				public Adapter EnclosingInstance
				{
					get
					{
						return enclosingInstance;
					}
					
				}
				public virtual string getMessage()
				{
					return "Illegal argument:" + e.Message;
				}
			}
			private void  InitBlock(InvokingComponentAdapterFactory enclosingInstance)
			{
				this.enclosingInstance = enclosingInstance;
			}
			private InvokingComponentAdapterFactory enclosingInstance;
			virtual public object InvocationResult
			{
				get
				{
					return invocationResult;
				}
				
			}
			public InvokingComponentAdapterFactory EnclosingInstance
			{
				get
				{
					return enclosingInstance;
				}
				
			}
			private System.Reflection.MethodInfo method;
			private object componentInstance = null;
			
			private object invocationResult;
			
			public Adapter(InvokingComponentAdapterFactory enclosingInstance, ComponentAdapter theDelegate):base(theDelegate)
			{
				InitBlock(enclosingInstance);
				try
				{
					method = theDelegate.ComponentImplementation.GetMethod(EnclosingInstance.methodName, EnclosingInstance.parameterTypes);
				}
				catch (System.MethodAccessException)
				{
					method = null;
				}
				catch (System.Security.SecurityException)
				{
					method = null;
				}
			}
			
			public override object GetComponentInstance(MutablePicoContainer picoContainer)
			{
				if (componentInstance == null)
				{
					componentInstance = base.GetComponentInstance(picoContainer);
					
					if (method != null)
					{
						try
						{
							invocationResult = method.Invoke(componentInstance, (object[]) EnclosingInstance.arguments);
						}
						catch (System.UnauthorizedAccessException e)
						{
							throw new PicoInvocationTargetInitializationException(e);
						}
						catch (System.ArgumentException e)
						{
							throw new AnonymousClassPicoInitializationException(e, this);
						}
						catch (System.Reflection.TargetInvocationException e)
						{
							throw new PicoInvocationTargetInitializationException(e.GetBaseException());
						}
					}
				}
				return componentInstance;
			}
		}
	}
}